inspirehep/migrator/tasks.py

```diff
@@ -126,54 +106,75 @@ def migrate_record_from_legacy(recid):
+
     if also_migrate is None:
         query = query.filter(LegacyRecordsMirror.valid.is_(None))
     elif also_migrate == "broken":
         query = query.filter(LegacyRecordsMirror.valid.isnot(True))
     elif also_migrate != "all":
         raise ValueError('"also_migrate" should be either None, "all" or "broken"')
 
-    if wait_for_results:
-        # if the wait_for_results is true we enable returning results from the
-        # migrate_recids_from_mirror task so that we could use them to
-        # synchronize migrate task (which in that case waits for the
-        # migrate_recids_from_mirror tasks to complete before it finishes).
-        tasks = []
-        migrate_recids_from_mirror.ignore_result = False
-
-    chunked_recids = chunker(res.recid for res in query.yield_per(CHUNK_SIZE))
-    for i, chunk in enumerate(chunked_recids):
-        scheduled_records = i * CHUNK_SIZE + len(chunk)
-        print(f"Scheduled {scheduled_records} records for migration")
-        if wait_for_results:
-            tasks.append(migrate_recids_from_mirror.s(chunk))
-        else:
-            migrate_recids_from_mirror.delay(chunk)
-
-    if wait_for_results:
-        job = group(tasks)
-        result = job.apply_async()
-        result.join()
-        migrate_recids_from_mirror.ignore_result = True
-        print("All migration tasks have been completed.")
+    recids_chunked = chunker(res.recid for res in query.yield_per(CHUNK_SIZE))
+    tasks = migrate_recids_from_mirror(
+        list(recids_chunked), disable_orcid_push=disable_orcid_push, wait=wait
+    )
+    print("All migration  tasks has been scheduled.")
+    return tasks
+
+
+# Set time_limit for task to 24 hours
+@shared_task(ignore_results=True, queue="migrator", acks_late=True)
+def migrate_recids_from_mirror(
+    recids_chunks, step_no=0, disable_orcid_push=True, wait=False
+):
+    """Task to migrate a chunked list of recids from the mirror."""
+    migration_steps = [
+        create_records_from_mirror_recids,
+        recalculate_citations,
+        index_record,
+    ]
+    if not disable_orcid_push:
+        migration_steps.append((run_orcid_push,))
+
+    if step_no >= len(migration_steps):
+        return
+
+    logger.info(f"Running migration step {step_no+1}/{len(migration_steps)}")
+
+    all_recs = 0
+    for row in recids_chunks:
+        all_recs += len(row)
+    if all_recs == 0:
+        logger.warning("There are no records to migrate. Terminating.")
+        return None
+    logger.info(f"Processing {all_recs} records in this migration step")
+    step = migration_steps[step_no]
+    header = (step.s(r) for r in recids_chunks)
+    chord_task = chord(header)(
+        migrate_recids_from_mirror.s(
+            step_no=step_no + 1, disable_orcid_push=disable_orcid_push, wait=True
+        )
+    )
+    if wait and chord_task:
+        with allow_join_result():
+            logger.info("Waiting for tasks to finish")
+            chord_task.get()
```

Comment:
this is a terrible way to do it, as now you're waiting within the task. Why are you doing this?

Reply:
I am aware of this but it's not possible to wait the way we talked yesterday. If you want we can discuss.

Reply:
yes, sorry, didn't want to be harsh. Thing is there's no point of using chords if we are waiting anyway. Let's discuss it.